dev/gencopy/gencopy.ml

```diff
@@ -0,0 +1,369 @@
+open Ppxlib
+
+module Ast_builder = Ast_builder.Make (struct
+  let loc = Location.none
+end)
+
+let loc txt = Ast_builder.Located.mk txt
+let ident_loc s = loc (Longident.parse s)
+
+module Ast_io = Ppxlib__.Utils.Ast_io
+
+module Copy = struct
+  type context = {
+    from : string; (* Module name of the origin AST *)
+    to_ : string; (* Module name of the destination AST *)
+    inside : string; (* Sub-module in Ast_x, e.g. Parsetree or Asttypes *)
+  }
+
+  let mk_ident ~root ~sub name =
+    ident_loc (Printf.sprintf "%s.%s.%s" root sub name)
+
+  let from_ident ~context name =
+    mk_ident ~root:context.from ~sub:context.inside name
+
+  let to_ident ~context name =
+    mk_ident ~root:context.to_ ~sub:context.inside name
+
+  let from_var_name i = Printf.sprintf "f%d" i
+  let to_var_name i = Printf.sprintf "g%d" i
+  let from_var i = Ast_builder.ptyp_var (from_var_name i)
+  let to_var i = Ast_builder.ptyp_var (to_var_name i)
+
+  let from_type ~context { ptype_name; ptype_params; _ } =
+    let longident = from_ident ~context ptype_name.txt in
+    let params = List.mapi (fun i _ -> from_var i) ptype_params in
+    Ast_builder.ptyp_constr longident params
+
+  let to_type ~context { ptype_name; ptype_params; _ } =
+    let longident = to_ident ~context ptype_name.txt in
+    let params = List.mapi (fun i _ -> to_var i) ptype_params in
+    Ast_builder.ptyp_constr longident params
+
+  let fun_name ~context ptype_name =
+    let name =
+      match ptype_name with
+      | "t" -> String.lowercase_ascii context.inside
+      | _ -> ptype_name
+    in
+    Printf.sprintf "copy_%s" name
+
+  let arrow_type args ret =
+    let rev_args = List.rev args in
+    let rec aux acc args =
```

Comment:
Could this be simplified to a `fold_left`: 

```ocaml
List.fold_left (fun acc hd -> Ast_builder.ptyp_arrow Nolabel hd acc) ret rev_args
```

Reply:
I can probably also remove the `List.rev` and make it a `List.fold_right` straight away!